src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,225 @@
+
+        public TResult Current
+        {
+            get
+            {
+                RawFindData findData = new RawFindData(_info, _currentPath, _originalFullPath, _originalUserPath);
+                return _transform(ref findData);
+            }
+        }
+
+        object IEnumerator.Current => Current;
+
+        public bool MoveNext()
+        {
+            if (_lastEntryFound)
+                return false;
+
+            RawFindData findData = default;
+            do
+            {
+                FindNextFile();
+                if (!_lastEntryFound && _info != null)
+                {
+                    // If needed, stash any subdirectories to process later
+                    if (_pending != null && (_info->FileAttributes & FileAttributes.Directory) != 0
+                        && !PathHelpers.IsDotOrDotDot(_info->FileName))
+                    {
+                        string subDirectory = PathHelpers.CombineNoChecks(_currentPath, _info->FileName);
+                        _pending.Enqueue((CreateDirectoryHandle(subDirectory), subDirectory));
+                    }
+
+                    findData = new RawFindData(_info, _currentPath, _originalFullPath, _originalUserPath);
+                }
+            } while (!_lastEntryFound && !_predicate(ref findData));
+
+            return !_lastEntryFound;
+        }
+
+        private unsafe void FindNextFile()
+        {
+            Interop.NtDll.FILE_FULL_DIR_INFORMATION* info = _info;
+            if (info != null && info->NextEntryOffset != 0)
+            {
+                // We're already in a buffer and have another entry
+                _info = (Interop.NtDll.FILE_FULL_DIR_INFORMATION*)((byte*)info + info->NextEntryOffset);
+                return;
+            }
+
+            // We need more data
+            if (GetData())
+                _info = (Interop.NtDll.FILE_FULL_DIR_INFORMATION*)_pinnedBuffer.AddrOfPinnedObject();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void NoMoreFiles()
+        {
+            _info = null;
+            if (_pending == null || _pending.Count == 0)
+            {
+                _lastEntryFound = true;
+            }
+            else
+            {
+                // Grab the next directory to parse
+                (IntPtr Handle, string Path) next = _pending.Dequeue();
+                Interop.Kernel32.CloseHandle(_directoryHandle);
+                _directoryHandle = next.Handle;
+                _currentPath = next.Path;
+                FindNextFile();
+            }
+        }
+
+        public void Reset()
+        {
+            throw new NotSupportedException();
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected void Dispose(bool disposing)
+        {
+            byte[] buffer = Interlocked.Exchange(ref _buffer, null);
```

Comment:
Is it correct to manipulate managed objects during finalization (disposing=false)? 

>If the method call comes from a finalizer (that is, if disposing is false), only the code that frees unmanaged resources executes. Because the order in which the garbage collector destroys managed objects during finalization is not defined, calling this Dispose overload with a value of false prevents the finalizer from trying to release managed resources that may have already been reclaimed.

So I'm not sure that queue is in a good state then.

Reply:
@jkotas can the member Queue<T> get collected before the finalizer runs?

Reply:
I am not sure what you are asking about. The general rules are:
- The GC can start running finalizers as soon as the object reference becomes unreachable. Note that JIT optimizations can make the object unrechable sooner than what you think. You typically need to have strategically placed GC.KeepAlive to ensure that the JIT optimizations will extend the lifetime as necessary.
- The order in which finalizers run is non-deterministic

Reply:
So is it safe to enumerate the Queue in this Dispose if it's called during finalization?

Reply:
Queue does not have any finalizers. It is fine to enumerate it even during finalization.

The risky part here is that the Queue contains unprotected unmanaged handles. You need to be careful about that no code is accessing these handles anymore while the finalizers run.